skcosmo/pcovr/kpcovr.py

```diff
@@ -0,0 +1,383 @@
+            Returns the instance itself.
+
+        """
+
+        X, Y = check_X_y(X, Y, y_numeric=True, multi_output=True)
+        self.X_fit_ = X.copy()
+
+        if self.n_components is None:
+            self.n_components = min(X.shape)
+
+        self._centerer = KernelFlexibleCenterer()
+        K = self._get_kernel(X)
+
+        if W is None:
+            if Yhat is None:
+                W = (np.linalg.pinv(K, rcond=self.regularization_) @ Y).reshape(
+                    X.shape[0], -1
+                )
+            else:
+                W = np.linalg.lstsq(K, Yhat, rcond=self.regularization_)[0]
+
+        if Yhat is None:
+            Yhat = K @ W
+
+        self._fit(K, Yhat, W)
+
+        self.ptk_ = self.pt__ @ K
+        self.pty_ = self.pt__ @ Y
+
+        if self.fit_inverse_transform:
+            self.ptx_ = self.pt__ @ X
+
+        self.pky_ = self.pkt_ @ self.pty_
+
+        self.components_ = self.pkt_.T  # for sklearn compatibility
+        return self
+
+    def predict(self, X=None, T=None):
+        """Predicts the property values"""
+
+        check_is_fitted(self, ["pky_", "pty_"])
+
+        if X is None and T is None:
+            raise ValueError("Either X or T must be supplied.")
+
+        if X is not None:
+            X = check_array(X)
+            K = self._get_kernel(X)
+            return K @ self.pky_
+        else:
+            T = check_array(T)
+            return T @ self.pty_
+
+    def transform(self, X):
+        """
+        Apply dimensionality reduction to X.
+
+        X is projected on the first principal components as determined by the
+        modified Kernel PCovR distances.
+
+        Parameters
+        ----------
+        X: array-like, shape (n_samples, n_features)
+            New data, where n_samples is the number of samples
+            and n_features is the number of features.
+
+        """
+
+        check_is_fitted(self, ["pkt_", "X_fit_"])
+
+        X = check_array(X)
+        K = self._get_kernel(X, self.X_fit_)
+        return K @ self.pkt_
+
+    def inverse_transform(self, T):
+        """Transform input data back to its original space.
+
+        .. math::
+
+            \\mathbf{\\hat{X}} = \\mathbf{T} \\mathbf{P}_{TX}
+                              = \\mathbf{K} \\mathbf{P}_{KT} \\mathbf{P}_{TX}
+
+
+        Similar to KPCA, the original features are not always recoverable,
+        as the projection is computed from the kernel features, not the original
+        features, and the mapping between the original and kernel features
+        is not one-to-one.
+
+        Parameters
+        ----------
+        T: array-like, shape (n_samples, n_components)
+            Projected data, where n_samples is the number of samples
+            and n_components is the number of components.
+
+        Returns
+        -------
+        X_original array-like, shape (n_samples, n_features)
+        """
+
+        return T @ self.ptx_
+
+    def score(self, X, Y):
```

Comment:
Do we want the "real" KPCovR loss in addition to (or in place of) the one here?

Reply:
One other comment since I've recently been looking into sklearn's scoring infrastructure. In sklearn, the `score` method for each estimator should only return a _single  value_, and this is required for compatibility with the existing pipeline and cross validation infrastructure. I suggest that we separate the (K)PCovR losses into separate methods, e.g., `kpcovr.lkpca` and `kpca.lkrr`, and then essentially make `kpcovr.score` a wrapper around these that returns the sum (or mean) of the two. Things would get a little more complex if we add additional arguments to `kpcovr.score` (e.g., throwing the Gram loss in the mix or a weighted mean of the losses), but I think it would still be usable in sklearn pipelines and cross validation.

Reply:
That's a solid point -- although I'm split on whether to use the _true_ KPCovR loss. On one hand, it _is_ the loss being optimized, but on the other hand, it is not necessarily the loss that people are interested in. What are your thoughts?